packages/allure-js-commons/src/AllureCommandStep.ts

```diff
@@ -0,0 +1,267 @@
+      });
+    }
+  }
+
+  label(label: string, value: string): void {
+    if (!this.metadata.labels) {
+      this.metadata.labels = [];
+    }
+
+    this.metadata.labels.push({
+      name: label,
+      value,
+    });
+  }
+
+  link(url: string, name?: string, type?: string): void {
+    if (!this.metadata.links) {
+      this.metadata.links = [];
+    }
+
+    this.metadata.links.push({
+      name,
+      url,
+      type,
+    });
+  }
+
+  parameter(name: string, value: string, options?: ParameterOptions): void {
+    if (!this.metadata.parameter) {
+      this.metadata.parameter = [];
+    }
+
+    this.metadata.parameter.push({
+      name,
+      value,
+      hidden: options?.hidden || false,
+      excluded: options?.excluded || false,
+    });
+  }
+
+  epic(epic: string) {
+    this.label(LabelName.EPIC, epic);
+  }
+
+  feature(feature: string) {
+    this.label(LabelName.FEATURE, feature);
+  }
+
+  story(story: string) {
+    this.label(LabelName.STORY, story);
+  }
+
+  suite(name: string) {
+    this.label(LabelName.SUITE, name);
+  }
+
+  parentSuite(name: string) {
+    this.label(LabelName.PARENT_SUITE, name);
+  }
+
+  subSuite(name: string) {
+    this.label(LabelName.SUB_SUITE, name);
+  }
+
+  owner(owner: string) {
+    this.label(LabelName.OWNER, owner);
+  }
+
+  severity(severity: string) {
+    this.label(LabelName.SEVERITY, severity);
+  }
+
+  tag(tag: string) {
+    this.label(LabelName.TAG, tag);
+  }
+
+  issue(name: string, url: string) {
+    this.link(url, name, LinkType.ISSUE);
+  }
+
+  tms(name: string, url: string) {
+    this.link(url, name, LinkType.TMS);
+  }
+
+  attach(source: string | Buffer, type: string): void {
+    this.attachments.push({
+      name: "attachment",
+      source: Buffer.isBuffer(source) ? source.toString("base64") : source,
+      type,
+    });
+  }
+
+  async step(name: string, body: StepBodyFunction): Promise<void> {
+    if (!this.metadata.steps) {
+      this.metadata.steps = [];
+    }
+
+    const nestedStep = new AllureCommandStepExecutable(name);
+    const {
+      labels = [],
+      links = [],
+      parameter = [],
+      categories = [],
+      environmentInfo = {},
+      steps = [],
+    } = await nestedStep.start(body);
+
+    this.metadata.labels = (this.metadata.labels || []).concat(labels);
+    this.metadata.links = (this.metadata.links || []).concat(links);
+    this.metadata.parameter = (this.metadata.parameter || []).concat(parameter);
+    this.metadata.categories = (this.metadata.categories || []).concat(categories);
+    this.metadata.environmentInfo = Object.assign(
+      this.metadata.environmentInfo || {},
+      environmentInfo,
+    );
+    this.metadata.steps = (this.metadata.steps || []).concat(steps);
+  }
+
+  async start(body: StepBodyFunction): Promise<AttachmentMetadata> {
+    const startDate = new Date().getTime();
+
+    try {
+      const res = body.call(this, this);
```

Comment:
this start will be executed on the reporter thread for the paralell execution??? if so something can broke

Reply:
It depends. For example `wdio` attaches separated report for each thread. 
Cucumberjs doesn't, but it executes steps inside world methods, which call in separated threads too.
What kind of doubt you have?